nrf/PWM: Use extmod/machine.pwm for PWM.

robert-hh opened the pull request:
Making it compatible with the other ports providing the machine.PWM class. This PR includes a hardware PWM variant for nrf52xxx based boards, and a software PWM for other boards. It is a breaking change in that it changes the existing machine.PWM API for this port, which however was not documented.

Enabled for most nrf52xxx boards and some nrf51xxx boards. Properties:

nrf52xxx hardware PWM: 
- 3 or 4  PWM modules with independent frequency. Each module has four channels with independent duty cycle and polarity. So in total 12 to 16 PWM signals can be created.
- Frequency range 4Hz - ~5 MHz (16MHz / 3).
- Duty cycle methods duty() with range 0-100, duty_u16() and duty_ns(). The duty() method is supported for legacy.
- Every Pin can be used with PWM.

nrf51xxx software PWM:
- Can be enabled on any Pin in the P0.x range.
- Frequency range 1Hz - 3906 Hz.
- Duty cycle methods duty() with range 0-100, duty_u16() and duty_ns(). The duty() method is supported for legacy.

All pins of the software PWM share the same frequency, but may have different duty cycle setting. Since the drivers/softpwm.c module is used for software PWM which is also used by the `music` class, playing music will change the PWM frequency.

Note: The first commit in this PR just adds checks for the input arguments to the existing implementation. That avoids a crash when invalid values are used.

Tested with the `seeed XIAO nrf52840` and `micro:bit` boards.

robert-hh committed:
nrf/modules/machine/pwm: Add paramter checks and error messages.

robert-hh committed:
nrf/modules/machine/pwm: Use extmod/machine_pwm.c for PWM module.

This is a breaking change, making the hardware PWM on the nrf port
compatible with the other ports providing machine.PWM.

Frequency range 4Hz - ~5.4 MHz.  The base clock range is 125kHz to 16 MHz,
and the divider range is 3 - 32767.

The hardware supports up to four outputs per PWM device with different duty
cycles, but only one output is (and was) supported.

robert-hh committed:
nrf/modules/machine/soft_pwm: Add PWM for nrf51x boards using soft PWM.

Using extmod/machine_pwm.c for the Python bindings and the existing
softpwm.c driver, by just adding the interface.

Properties:
- Frequency range 1-3906 Hz.
- All PWM outputs run at the same frequency but can have different duty
  cycles.
- Limited to the P0.x pins.

Since it uses the existing softpwm.c mechanism, it will be affected by
playing music with the music class.

dpgeorge commented on ports/nrf/boards/arduino_nano_33_ble_sense/mpconfigboard.h:
```diff
@@ -15,6 +15,11 @@
+#define MICROPY_PY_MACHINE_PWM_INIT     (1)
+#define MICROPY_PY_MACHINE_PWM_DUTY     (1)
+#define MICROPY_PY_MACHINE_PWM_DUTY_U16_NS (1)
+#define MICROPY_PY_MACHINE_PWM_INCLUDEFILE "ports/nrf/modules/machine/pwm.c"
```
I don't think it makes sense to repeat these 5 lines in all mpconfigboard.h files.  Instead, the 4 latter ones (_INIT, _DUTY, _DUTY_U16_NS, _INCLUDEFILE) can all go in the common mpconfigport.h file.  And then `MICROPY_PY_MACHINE_PWM` can either stay here in the board specific config file, or default to enabled but with a `#ifndef ...` guard (so a board can disable it if needed).

robert-hh replied:
That's a good suggestion. I prefer to have `MICROPY_PY_MACHINE_PWM` in the board files.

robert-hh committed:
nrf/modules/machine/pwm: Support using all 4 channels of a PWM module.

These have the same frequency, but can have different duty cycle and
polarity.

pwm.deinit() stops all channels of a module, but does not release the
module.  pwm.init() without arguments restarts all outputs.

dpgeorge commented:
It looks like the file `pwm_old.c` is not used.  In that case it should be deleted.  If anyone needs the old behaviour/code they can use git to go back in the history.

robert-hh commented:
Of course. It was not meant to be there. Maybe I used `git mv` instead of just `mv`.

dpgeorge commented:
Thanks for this, it's a very welcome improvement!

Squashed and cleaned up the commits and merged.

robert-hh commented:
Thank you for merging. That's part of the attempt to harmonize PWM. Still some discussions around it are open.
